### PR TITLE
fix(compiler): using `print()` in bad places causes panics

### LINKED ISSUE #1062: using `print()` in bad places causes compiler panics
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

I tried writing some invalid code:

Example 1:
```
print("hey").get("x");
```

Example 2:
```
let x = "my name is ${print("chris")}";
```

Example 3:
```
let x = 5 + print("hello");
```

### I expected this:

I expected `wing compile` to show me relevant diagnostics, like `Expression must be a class or resource instance to access property "get", instead found type "void"`.

### Instead, this happened:

Instead, when I was trying this in VS Code my language server crashed due to compiler panics. All of the examples panic with an error like:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', libs/wingc/src/type_check.rs:1691:37
```

### Component:

Compiler

### Wing Version:

0.4.59

### Node.js Version:

19.2.0

### Platform(s):

MacOS

### Anything else?

I think the issue is related to how the compiler models return types as `Option<Type>` / `Option<TypeRef>`. When `type_check_exp` is called on an expression like `print("hey")`, the void type is modeled as `Option::None`, so `None` is the value that is returned. When other parts of the type checker calls `unwrap()` on the result of `type_check_exp`, they all panic.

Perhaps we could model a `Void` option inside the `Type` enum?

### PR DESCRIPTION
**Pull request by Chriscbr**
Fixes #1062

The implementation adds a new type `void` to our type system, used to represent the return type of functions that do not return a value. This type is not user-constructible (you cannot create a `void` value) and the `void` type is only equal to itself during type checking. In doing so we're able to remove a _lot_ of `unwrap` calls that corresponded to compiler panics. I've included a few examples in the `print_invalid` unit test.

*By submitting this pull request, I confirm that my contribution is made under the terms of the 
[Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: examples/tests/invalid/print_invalid.w
Comment (eladb) [inline examples/tests/invalid/print_invalid.w:1]: wonderful!

**Review by eladb — COMMENTED** [review]
Looks great to me

**Review by MarkMcCulloh — APPROVED** [review]

File: libs/wingc/src/type_check/symbol_env.rs
Comment (Chriscbr) [inline libs/wingc/src/type_check/symbol_env.rs:13]: I feel like this might need some further modeling -- in some cases a SymbolEnv corresponds to an assignable block of code, like a function body (so we use this `return_type` field for type checking) but in other cases it doesn't correspond to a block of code (e.g. the environment inside a class definition). For now I inserted `void` where we previously used `None` in dummy `SymbolEnv` instances

Comment (MarkMcCulloh) [reply libs/wingc/src/type_check/symbol_env.rs:13]: I don't think your change made the ambiguity any worse, but I would definitely agree it needs fourth modeling

**Comment by Chriscbr** [comment]
I'll wait for @yoav-steinberg's blessing - added a `pr/do-not-merge` label in the mean time.

File: examples/tests/invalid/print_invalid.w
Comment (yoav-steinberg) [inline examples/tests/invalid/print_invalid.w:8]: Maybe change the test name to something like "can't use void in expressions" and add another test that verifies the same thing with a custom function (not `print`):
```wing
returns_nothing = () {};
let x = returns_nothing();
```

**Review by yoav-steinberg — APPROVED** [review]
I love this! I actually had a branch doing almost exactly the same change which I never used.

File: libs/wingc/src/type_check.rs
Comment (Chriscbr) [reply libs/wingc/src/type_check.rs:598]: Ah, I think I just temporarily added this to force a type error while refactoring this.. will remove (nice catch)

File: examples/tests/invalid/print_invalid.w
Comment (Chriscbr) [reply examples/tests/invalid/print_invalid.w:8]: Good idea

Comment (Chriscbr) [reply examples/tests/invalid/print_invalid.w]: BTW, found a bug, as it was possible to assign void to a variable - fixed and added it to the example.

**Pull request merged**
